[14.0][IMP] product supplierinfo intercompany[_multi_company]: refactor, archiving

HekkiMelody opened the pull request:
These commits

- slightly refactor these two modules so that the method doesn't need to be overwritten
- implement deletion and recreation of supplierinfo when archiving and unarchiving product templates and product pricelists

More details in each commit

Kev-Roche committed:
[IMP] product_supplierinfo_intercompany: ensure coherent currency

Ensure pricelist and supplierinfo use the same currency

giarve committed:
[FIX] product_supplierinfo_intercompany: Dynamic intercompany pricelist

Dynamic intercompany pricelist with cost / list price as item

HekkiMelody committed:
[IMP] product_supplierinfo_intercompany: info creation/deletion

The module has been refactored to ease inheritance

When archiving and unarchiving a product template,
delete the automatically-created intercompany supplierinfo records
for that product

When archiving or unarchiving an intercompany pricelist, delete or
recreate all the intercompany supplierinfo

HekkiMelody committed:
[REF] product_supplierinfo_intercompany_multi_company: info creation

Use inheritance instead of overwriting methods in
product.intercompany.supplier.mixin

Do not create supplierinfo if the product is only visible to
the same company as the pricelist

giarve reviewed (approved):
for _condition_supplierinfo_create_or_update

can you also add these fields such that when they are changed the pricelist recomputes?

this is when it is based on cost field and similars

https://github.com/OCA/multi-company/pull/821/files#diff-19f4129027b1cecdf793d8eb674abc4a90b136c629ee2b4cae958ff1a5f8d2f9R75

HekkiMelody commented:
> for _condition_supplierinfo_create_or_update
> 
> can you also add these fields such that when they are changed the pricelist recomputes?
> 
> this is when it is based on cost field and similars
> 
> https://github.com/OCA/multi-company/pull/821/files#diff-19f4129027b1cecdf793d8eb674abc4a90b136c629ee2b4cae958ff1a5f8d2f9R75

Thank you for your feedback!

I cherry-picked your commit as well as a previous one about currency. The modules are quite different, and doing more is definitely outside the scope of this PR.

I realized after I had already started that the change you requested wasn't strictly necessary because in v14 pricelists are recomputed ALWAYS, there was no list of fields to check against, and therefore there was no need to add any fields to that list. Still, while I was at it, I added it. Let me know what you think.

giarve commented:
> > for _condition_supplierinfo_create_or_update
> > can you also add these fields such that when they are changed the pricelist recomputes?
> > this is when it is based on cost field and similars
> > https://github.com/OCA/multi-company/pull/821/files#diff-19f4129027b1cecdf793d8eb674abc4a90b136c629ee2b4cae958ff1a5f8d2f9R75
> 
> Thank you for your feedback!
> 
> I cherry-picked your commit as well as a previous one about currency. The modules are quite different, and doing more is definitely outside the scope of this PR.
> 
> I realized after I had already started that the change you requested wasn't strictly necessary because in v14 pricelists are recomputed ALWAYS, there was no list of fields to check against, and therefore there was no need to add any fields to that list. Still, while I was at it, I added it. Let me know what you think.

great. For me it is interesting because this module is having lots of changes in between versions but nobody is set as maintainer so not guiding the boat and it was just to make sure we are following the right approach so when you migrate to v16 for instance you get the same features and non breaking changes

SirPyTech commented on product_supplierinfo_intercompany/models/product_intercompany_supplier_mixin.py:
```diff
@@ -26,6 +26,7 @@ def _prepare_intercompany_supplier_info(self, pricelist):
         }
         return res
 
+    @api.model
```
**todo**: This method has a `for record in self` loop so it depends on the records in `self`, therefore it should not be `api.model`.

SirPyTech replied:
:+1:

SirPyTech commented on product_supplierinfo_intercompany/models/product_intercompany_supplier_mixin.py:
```diff
@@ -43,15 +44,31 @@ def _synchronise_supplier_info(self, pricelists=None):
+    def _synchronise_supplier_info_for_record(self, record, pricelist, supplierinfo):
+        if self._condition_supplierinfo_create_or_update(
+            record, pricelist, supplierinfo
+        ):
```
**question**: Here and below, why are we calling on `record` methods that are also defined for `self`?
For instance, I think:
```suggestion
    def _synchronise_supplier_info_single(self, pricelist, supplierinfo):
        self.ensure_one()
        if self._condition_supplierinfo_create_or_update(
            pricelist, supplierinfo
        ):
```
would be more intuitive, but maybe it's just me, what do you think?

SirPyTech replied:
👍

SirPyTech commented on product_supplierinfo_intercompany_multi_company/models/product_intercompany_supplier_mixin.py:
```diff
@@ -1,50 +1,26 @@
+        res = super()._synchronise_supplier_info_for_record(
+            record, pricelist, supplierinfo
+        )
+        if (
+            record._has_intercompany_price(pricelist)
+            and record.company_id == pricelist.company_id
+        ):
```
**suggestion**: We could add this condition to the `_condition_..._unlink` method instead: that would use inheritance and should give the same result, what do you think?

HekkiMelody replied:
I've checked, but it's not quite the same. If I add it to `_condition_..._unlink` but `_condition_..._create_or_update` is true, it will not be executed. This way, it's executed regardless (meaning a supplierinfo record could be created and then immediately deleted, in some cases). I've kept it as is to preserve the existing logic since it's a refactor.

SirPyTech replied:
Right, that's because of the **el**if, thanks for the clarification 👍

SirPyTech commented on product_supplierinfo_intercompany/models/product_pricelist.py:
```diff
@@ -43,3 +43,14 @@ def _unactive_intercompany(self):
             "generated_supplierinfo_ids"
         ).unlink()
+
+    def write(self, vals_list):
```
**typo**: Should be `vals`

SirPyTech replied:
👍

SirPyTech commented on product_supplierinfo_intercompany/models/product_pricelist.py:
```diff
@@ -43,3 +43,14 @@ def _unactive_intercompany(self):
+                    if rec.active:
+                        rec._active_intercompany()
+                    else:
+                        rec._unactive_intercompany()
```
**suggestion**: Can we batch records to activate/unactivate and then call the corresponding method on the batch? That could speed up the process

SirPyTech replied:
👍

SirPyTech commented on product_supplierinfo_intercompany/models/product_intercompany_supplier_mixin.py:
```diff
@@ -17,13 +17,19 @@ def _get_intercompany_supplier_info_domain(self, pricelist):
+
+        # Set context to ensure we are taking the list price / cost / etc. fields
+        # from the company set in the pricelist as it is a company-dependent field!
+        product_with_context = self.with_company(pricelist.company_id.id)
```
**question**: Could you please clarify which one is the company-dependent field that needs the right company to be set?

HekkiMelody replied:
I thought it would be `price` but looking at the compute method, it doesn't seem to be the case. Maybe a difference with v16.

I've deleted the comment and reverted the change

SirPyTech replied:
👍

SirPyTech commented on product_supplierinfo_intercompany/models/product_intercompany_supplier_mixin.py:
```diff
@@ -17,13 +17,19 @@ def _get_intercompany_supplier_info_domain(self, pricelist):
+        # from the company set in the pricelist as it is a company-dependent field!
+        product_with_context = self.with_company(pricelist.company_id.id)
+
+        price = product_with_context.uom_id._compute_price(self.price, self.uom_po_id)
```
**suggestion**: `product.pricelist.get_product_price` seems to be the corresponding method to `product.pricelist._get_product_price` used in `16.0`; `uom_id`'s method simply converts from one UoM to another.

HekkiMelody replied:
I kept the computation as it was previously in v14

SirPyTech replied:
👍

SirPyTech commented on product_supplierinfo_intercompany/models/product_pricelist.py:
```diff
@@ -12,6 +12,10 @@ class ProductPricelist(models.Model):
         default=False, inverse="_inverse_intercompany_supplier"
     )
 
+    intercompany_supplier_lead_time = fields.Float(
```
**chore**: I don't see how adding this field helps with `[FIX] Dynamic intercompany pricelist with cost / list price as item`, can we move it to another more appropriate commit?

SirPyTech replied:
👍

SirPyTech commented on product_supplierinfo_intercompany/models/product_template.py:
```diff
@@ -65,9 +65,18 @@ def create(self, vals):
+            or "standard_price" in vals
+            or "company_id" in vals
+            or "company_ids" in vals
+            or "active" in vals
```
**suggestion**: Please move this list of fields to its own method so it can be overridden if other modules add fields that need to `update_intercompany_prices`.

Also, how did we choose these fields? Can they be retrieved in some more dynamic way, if that's not overkill?

HekkiMelody replied:
I'm not entirely sure, as I've simply taken these fields from the commit by @giarve which I've cherry-picked here

Original: https://github.com/OCA/multi-company/pull/821/commits/78a47255fb145659790db01df444dcb3a7ca783e#diff-19f4129027b1cecdf793d8eb674abc4a90b136c629ee2b4cae958ff1a5f8d2f9R75-R82

SirPyTech replied:
👍

SirPyTech commented on product_supplierinfo_intercompany/models/product_intercompany_supplier_mixin.py:
```diff
@@ -17,13 +17,19 @@ def _get_intercompany_supplier_info_domain(self, pricelist):
 
```
**chore**: Please follow https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#71commit-message for the commit message, right now it is missing the module name: 
<img width="716" height="121" alt="image" src="https://github.com/user-attachments/assets/cd80e128-6cc0-4de8-ab2d-7b2162949fcf" />

SirPyTech replied:
👍

SirPyTech commented on product_supplierinfo_intercompany/models/product_intercompany_supplier_mixin.py:
```diff
@@ -23,6 +23,7 @@ def _prepare_intercompany_supplier_info(self, pricelist):
             "name": pricelist.company_id.partner_id.id,
```
**chore**: Please follow https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#71commit-message for the commit message, right now it is missing the module name: <img width="716" height="121" alt="image" src="https://github.com/user-attachments/assets/5cfb746e-b9a2-477b-8ab0-c7ba09a30bbc" />

SirPyTech replied:
👍

SirPyTech commented on product_supplierinfo_intercompany/models/product_intercompany_supplier_mixin.py:
```diff
@@ -17,13 +17,19 @@ def _get_intercompany_supplier_info_domain(self, pricelist):
+
+        # Set context to ensure we are taking the list price / cost / etc. fields
+        # from the company set in the pricelist as it is a company-dependent field!
+        product_with_context = self.with_company(pricelist.company_id.id)
```
**suggestion**: The tests are already well structured for multi-company testing, is it possible to add a test for this failing use-case?

HekkiMelody replied:
I've reverted the change since the company doesn't actually seem to affect the computation of the `price` field in v14

SirPyTech replied:
👍

SirPyTech reviewed (changes requested):
Thanks for the PR!
I only reviewed the code and seems mostly good to me, big thanks for keeping a clean commit history 👏.
I have some notes/suggestions, let me know what you think.

I am not a fan of cherry-picking commits from other open PRs because it doubles the developer (and reviewer's) work to sync changes in multiple PRs, but I understand that on the other hand it increases the probability to have a commit merged in _some PR_; in the end, it's a trade-off 🤷

HekkiMelody committed:
[IMP] product_supplierinfo_intercompany: lead time, inheritance

add intercompany lead time

Get list of fields that trigger a recomputation of the pricelist for
a product from a method to ease inheritance in custom modules

HekkiMelody commented:
Thank you for the review @SirPyTech ! ~~Unless I missed something~~, I have addressed all of your comments (EDIT: I missed one comment, I'm working on it), including adding two new commits to

- add intercompany lead time in a dedicated commit
- refactor the list of fields into a method for easier overriding as you suggested

Let me know what you think!

SirPyTech commented on product_supplierinfo_intercompany/models/product_intercompany_supplier_mixin.py:
```diff
@@ -62,6 +62,7 @@ def _condition_supplierinfo_create_or_update(self, pricelist, supplierinfo):
             self._has_intercompany_price(pricelist)
             and self.sale_ok
             and self.purchase_ok
+            and self.active
```
**thought**: A condition on `self.active` could also be added in `_condition_supplierinfo_unlink` since we want to delete the supplierinfo when the product is archived, what do you think?

HekkiMelody replied:
To be honest, as it is structured now, with the `elif`, I don't really see much use for `_condition_supplierinfo_unlink`, because effectively, if `_condition_supplierinfo_create_or_update` is True, you want to keep the supplierinfo, otherwise you want to delete it. There aren't any cases I can think of where you would want to keep around a supplierinfo that you don't want to update. That would be... obsolete?

But I also didn't want to add yet another change to the structure of this module, this PR is already too large as it is.

So maybe I could remove `_condition_supplierinfo_unlink`?

SirPyTech replied:
I understand that you either want to keep it or trash it, but keeping it does not mean that you have to update it.

I agree that as it is now, `_condition_supplierinfo_unlink` is not useful.
But since we really want to delete the `supplierinfo` of archived products (that's the purpose of the commit), I think the product being archived should be explicitly a part of the condition for the supplierinfo to be deleted (i.e. be part of `_condition_supplierinfo_unlink`).

Consider this snippet where the supplierinfo might have to be just kept, and not updated nor deleted:
```py
if supplierinfo_needs_update:
    supplierinfo.write(...)
elif supplierinfo_needs_unlink:
    supplierinfo.unlink()
```
if the supplierinfo does not need to be updated (because its values are fine already, for instance), then we would end up deleting it.

HekkiMelody replied:
Added `or (not self.active)` to `_condition_supplierinfo_unlink` to ensure deletion of supplierinfo related to products that have been archived

SirPyTech replied:
And in the right commit of course, thanks!

SirPyTech commented on product_supplierinfo_intercompany/tests/test_supplier_intercompany.py:
```diff
@@ -347,3 +350,9 @@ def test_getting_supplier_with_sudo(self):
         self.assertEqual(supplier.name, partner)
+
+    def test_currency_consistency(self):
+        self.pricelist_intercompany.currency_id = self.env.ref("base.USD")
```
**suggestion**: This test passes even without the related change in the module, probably because this currency is the default one (see https://github.com/odoo/odoo/blob/cc0060e889603eb2e47fa44a8a22a70d7d784185/addons/product/models/product.py#L774)

Please use another currency, and maybe also `assert` it's not the default one.

SirPyTech replied:
Now the test fails without the related change, thanks!

SirPyTech reviewed (changes requested):
Thanks for all the changes!
Just a note to make one of the added tests really useful and some food for thought, the rest looks good

SirPyTech commented on product_supplierinfo_intercompany/models/product_intercompany_supplier_mixin.py:
```diff
@@ -62,8 +62,9 @@ def _condition_supplierinfo_create_or_update(self, pricelist, supplierinfo):
     def _condition_supplierinfo_unlink(self, pricelist, supplierinfo):
         self.ensure_one()
-        return bool(supplierinfo)
+        return bool(supplierinfo) or (not self.active)
```
**thought**: This `bool(supplierinfo)` smells a bit in my opinion because it is saying that we want to delete the `supplierinfo` every time it is _truthy_, i.e. it exists.

This should always be `True`, but in the rare (if not impossible) case when it isn't True but `self` is not `active`, then we want to delete this not-existing `supplierinfo` 😵‍💫.

I understand I asked for this in https://github.com/OCA/multi-company/pull/859/changes/516b472ac6648dac1d767bf85f48d61716cd863a#r2606687413, and I am still quite sure the `active` condition should be here for the reasons explained in that thread, but I wouldn't know how to make it more readable; I think I'm missing something.
Maybe `(supplierinfo and not self.active) or supplierinfo`? But that would make the `self.active` condition redundant.

Anyway, `unlink`ing an empty recordset can be done and simply does nothing, so this isn't an issue.

francesco-ooops reviewed (approved):
Functional ok!

@OCA/intercompany-maintainers can this be merged please? thanks!

francesco-ooops commented:
@OCA/intercompany-maintainers ping :)

ivs-cetmix commented:
Hi @HekkiMelody , thank you for your contribution! Shouldn't some non-cherry-picked commits be squashed before we merge this PR?

HekkiMelody commented:
Hi @ivs-cetmix thank you for your time.

I personally think having small commits each with a single logical change makes it easier to both review the PR and review the history of the module via git later on, so I would prefer to keep them if possible, since otherwise there would only a couple very very large commits with lots of unrelated changes.

However, I'm happy to squash my commits if it's necessary for the merge. Please let me know your view on it. Thanks!

ivs-cetmix commented:
> I personally think having small commits each with a single logical change makes it easier to both review the PR and review the history of the module via git later on, so I would prefer to keep them if possible, since otherwise there would only a couple very very large commits with lots of unrelated changes.
> 
> However, I'm happy to squash my commits if it's necessary for the merge. Please let me know your view on it. Thanks!

@HekkiMelody I totally agree with you as I would also prefer to have changes in separate commits. However this is not about my personal preferecnes, this is about following the OCA rules and recommendations which require commits to be squashed prior to merge. 
If you can squash some of them while also updating the commit message - this would be perfect. Not mandatory to squash each and every though, and of cause cherry-picked commits should remain as is.

HekkiMelody commented:
> @HekkiMelody I totally agree with you as I would also prefer to have changes in separate commits. However this is not about my personal preferecnes, this is about following the OCA rules and recommendations which require commits to be squashed prior to merge. If you can squash some of them while also updating the commit message - this would be perfect. Not mandatory to squash each and every though, and of cause cherry-picked commits should remain as is.

I have squashed however many commits I could (while keeping one commit per module) because I want this PR merged and not open a debate.

But I'd like to note that I can't find any rule or recommendation "which require commits to be squashed prior to merge."

https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#7git

> Only make a single commit per logical change set.

It was also recently discussed in the contributor mailing list, for example here: https://odoo-community.org/groups/contributors-15/contributors-2164121?mode=thread&date_begin=&date_end=

And I think this squashing makes the PR actively worse, since the commits were curated and grouped by logical change.

ivs-cetmix commented:
@HekkiMelody thank you! Once again - I totally share your approach however there are rules we all need to follow. 
/ocabot merge patch
